### ISSUE TITLE
Capture context rich error message

### ISSUE THREAD

**Issue by sacOO7** [ably/ably-java#621]
- Current ErrorInfo class just captures the exception messages instead of the whole error context.
- Exception should be fully wrapped while being consumed.
- It shouldn't leave any missing trail behind for captures exceptions.
Link to read about exception handling practices - https://stackify.com/best-practices-exceptions-java/

**Comment by sacOO7** [comment]
Proposed solution 
- Extend ```ErroInfo``` with native ```Exception``` class.
(as stated, ```ErrorInfo``` is an exception type encapsulating error information containing an Ably-specific error code and generic status code, but it's not a java exception ... yet)
- Pass original exception as a **cause** while creating new ```ErrorInfo``` exception.
- Include exception **cause** while logging ErrorInfo object.

**Comment by mattheworiordan** [comment]
I suppose this proposal, this is what I have done in Ruby, see https://github.com/ably/ably-ruby/blob/main/lib/ably/models/error_info.rb#L33-L57 and https://github.com/ably/ably-ruby/blob/main/lib/ably/exceptions.rb#L16

**Comment by QuintinWillison** [comment]
We already have an `AblyException` class which encapsulates an `ErrorInfo` instance, which is an architectural choice - using composition - that had already been made (rightly or wrongly) in the past.

Modifying the inheritance tree above `ErrorInfo` (essentially widening its public interface) feels like a hammer to crack a nut, and is not the right place in the stack to do this.

There is already a [well-known, idiomatic constructor](https://docs.oracle.com/javase/7/docs/api/java/lang/Exception.html#Exception(java.lang.String,%20java.lang.Throwable)) on the base `Exception` class allowing us to pass a cause. Therefore this requirement may be solved by simply adding a constructor to our `AblyException` class that calls that.

Also, @sacOO7, can you please provide a breadcrumb link back to the original discussion that prompted you to create this issue so those passing by in future can gain more context as to the requirement. Thanks.

**Comment by sacOO7** [comment]
Yes, I think we already have ```AblyException`` class. I was looking into the use cases where we use ```ErrorInfo`` class and ```AblyException``` derived from ```ErrorInfo```. It seems maybe at some of the places, we are not using ```AblyException``` class effectively (Like the above case). I was also looking into the way exceptions are handled.
https://github.com/ably/ably-java/blob/647575906f716598174f0fac3ee6a9fbb7f805d5/lib/src/main/java/io/ably/lib/realtime/ChannelBase.java#L279
and 
https://github.com/ably/ably-java/blob/647575906f716598174f0fac3ee6a9fbb7f805d5/lib/src/main/java/io/ably/lib/realtime/ChannelBase.java#L289

Generally, it's not recommended to catch throwables
```
Throwable is the superclass of all exceptions and errors.  In a catch clause, it will not only catch all exceptions; it will also catch all errors. Errors are thrown by the JVM to indicate serious problems that are not intended to be handled by an application.
Typical examples for that are the OutOfMemoryError or the StackOverflowError. Both are caused by situations that are outside of the control of the application and can’t be handled.
```
I am still digging more to find where and how

**Comment by sacOO7** [comment]
So, I came up with the following findings after doing a deep analysis of current error handling implementation
Observations are based on the following guidelines
1. Custom error object shouldn't leave any missing trail of actual error source
- It becomes difficult while debugging the exact source of the issue

2. Code should not contain empty catch blocks.
- Error caught is not clear when it's caught without logging the problem

3. Every exception should be logged using logger instead of System.err or println
4. Be specific about catching the exception
5. Throwables shouldn't be caught

It seems we have code snippets that violate the above guidelines. Exceptions are supposed to be caught and well handled in order to get a clear idea about working/ non-working application behaviour

**Comment by sacOO7** [comment]
https://github.com/ably/ably-java/blob/647575906f716598174f0fac3ee6a9fbb7f805d5/lib/src/main/java/io/ably/lib/transport/WebSocketTransport.java#L226

- ErrorInfo is initiated without passing a complete exception context. It just takes message out of exception and original exception cause is lost.

**Comment by sacOO7** [comment]
https://github.com/ably/ably-java/blob/647575906f716598174f0fac3ee6a9fbb7f805d5/lib/src/main/java/io/ably/lib/types/ErrorInfo.java#L141

- When errorInfo is initialized, the original exception cause is lost. It just created an ErrorInfo object from the exception message.

**Comment by sacOO7** [comment]
https://github.com/ably/ably-java/blob/647575906f716598174f0fac3ee6a9fbb7f805d5/lib/src/main/java/io/ably/lib/http/HttpCore.java#L301

- System.err is used instead of a log handler to handle the exception.

**Comment by sacOO7** [comment]
https://github.com/ably/ably-java/blob/647575906f716598174f0fac3ee6a9fbb7f805d5/lib/src/main/java/io/ably/lib/realtime/ChannelStateListener.java#L62

https://github.com/ably/ably-java/blob/647575906f716598174f0fac3ee6a9fbb7f805d5/lib/src/main/java/io/ably/lib/types/MessageSerializer.java#L74

https://github.com/ably/ably-java/blob/647575906f716598174f0fac3ee6a9fbb7f805d5/lib/src/main/java/io/ably/lib/realtime/Presence.java#L362

- Exceptions are not properly logged.

**Comment by sacOO7** [comment]
https://github.com/ably/ably-java/blob/647575906f716598174f0fac3ee6a9fbb7f805d5/lib/src/main/java/io/ably/lib/http/Http.java#L49

- AblyException is caught, but ErrorInfo is passed to the callback method, so the original exception context is lost.

**Comment by sacOO7** [comment]
- Also, at some places, the only error message is logged rather than the whole exception enriched with cause and stacktrace
- Original exception context is lost in such cases.

**Comment by sacOO7** [comment]
- As per https://ably-real-time.slack.com/archives/C019LDWM9UH/p1602442482002500, we were not able to find the original exception causing the error.
- At places where an exception is thrown, `AblyException` should be preferred over `ErrorInfo` to preserve the cause of the exception.

**Comment by QuintinWillison** [comment]
Thanks for looking into this, @sacOO7. As [explained elsewhere (internal)](https://ably-real-time.slack.com/archives/C019LDWM9UH/p1603099010002500), this is not high priority now so is going to remain in the backlog until some of the other higher priority issues have been completed.
